### PR TITLE
[SPARK-15431][SQL] Support LIST FILE(s)|JAR(s) command natively

### PR DESCRIPTION
**Pull request by xwu0226**
## What changes were proposed in this pull request?

Currently command `ADD FILE|JAR <filepath | jarpath>` is supported natively in SparkSQL. However, when this command is run, the file/jar is added to the resources that can not be looked up by `LIST FILE(s)|JAR(s)` command because the `LIST` command is passed to Hive command processor in Spark-SQL or simply not supported in Spark-shell. There is no way users can find out what files/jars are added to the spark context.
Refer to [Hive commands](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Cli)

This PR is to support following commands:
`LIST (FILE[s] [filepath ...] | JAR[s] [jarfile ...])`
##### LIST FILE(s)

```
scala> spark.sql("add file hdfs://bdavm009.svl.ibm.com:8020/tmp/test.txt")
res1: org.apache.spark.sql.DataFrame = []
scala> spark.sql("add file hdfs://bdavm009.svl.ibm.com:8020/tmp/test1.txt")
res2: org.apache.spark.sql.DataFrame = []

scala> spark.sql("list file hdfs://bdavm009.svl.ibm.com:8020/tmp/test1.txt").show(false)
+----------------------------------------------+
|result                                        |
+----------------------------------------------+
|hdfs://bdavm009.svl.ibm.com:8020/tmp/test1.txt|
+----------------------------------------------+

scala> spark.sql("list files").show(false)
+----------------------------------------------+
|result                                        |
+----------------------------------------------+
|hdfs://bdavm009.svl.ibm.com:8020/tmp/test1.txt|
|hdfs://bdavm009.svl.ibm.com:8020/tmp/test.txt |
+----------------------------------------------+
```
##### LIST JAR(s)

```
scala> spark.sql("add jar /Users/xinwu/spark/core/src/test/resources/TestUDTF.jar")
res9: org.apache.spark.sql.DataFrame = [result: int]

scala> spark.sql("list jar TestUDTF.jar").show(false)
+---------------------------------------------+
|result                                       |
+---------------------------------------------+
|spark://192.168.1.234:50131/jars/TestUDTF.jar|
+---------------------------------------------+


scala> spark.sql("list jars").show(false)
+---------------------------------------------+
|result                                       |
+---------------------------------------------+
|spark://192.168.1.234:50131/jars/TestUDTF.jar|
+---------------------------------------------+
```
## How was this patch tested?

New test cases are added for Spark-SQL, Spark-Shell and SparkContext API code path.

### CODE REVIEW

**Comment by xwu0226** [comment]
cc @yhuai @hvanhovell @gatorsmile Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 1]
```diff
@@ -46,3 +46,33 @@ case class AddFile(path: String) extends RunnableCommand {
     Seq.empty[Row]
   }
 }
+
+/**
+ * Return a list of file paths that are added to resources.
+ * If file paths are provided, return the ones that are added to resources.
+ */
+case class ListFiles(files: Seq[String] = Seq.empty[String]) extends RunnableCommand {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: ListFilesCommand

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 2]
```diff
@@ -46,3 +46,33 @@ case class AddFile(path: String) extends RunnableCommand {
...
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    sparkSession.sparkContext.listFiles(files).map(Row(_))
+  }
+}
+
+/**
+ * Return a list of jar files that are added to resources.
+ * If jar files are provided, return the ones that are added to resources.
+ */
+case class ListJars(jars: Seq[String] = Seq.empty[String]) extends RunnableCommand {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: ListJarsCommand

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -1387,6 +1387,27 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   }
 
   /**
+    * Return a list of file paths that are added to resources.
+    * If file paths are provided, return the ones that are added to resources.
+    */
+  def listFiles(files: Seq[String] = Seq.empty[String]): Seq[String] = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: i think this one should not take any parameter, and if you need filtering, just do it in ListFilesCommand

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 4]
```diff
@@ -1724,6 +1745,22 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     postEnvironmentUpdate()
   }
 
+  /**
+    * Return a list of jar files that are added to resources.
+    * If jar files are provided, return the ones that are added to resources.
+    */
+  def listJars(jars: Seq[String] = Seq.empty[String]): Seq[String] = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 1 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: @rxin Thank you very much! I will make the change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 2 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Will change!

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 3 above)
Comment (xwu0226) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Agree. Thanks!

Comment (xwu0226) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: @rxin Just one concern about thiss one. It is possible that users just invoked listFiles or listJars directly with sparkContext. Do we want to provide filtering for this case? Right now, I have a [test case](https://github.com/xwu0226/spark/blob/21b092ab84b22abec93fde1fc1ca177db68d9f0d/core/src/test/scala/org/apache/spark/SparkContextSuite.scala#L159-L176) that covers this case.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: no they can filter themselves easily.

Comment (xwu0226) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: @rxin OK. Thanks!

**Comment by xwu0226** [comment]
@rxin Updated the code. Please help double check. Thank you very much!

**Comment by cloud-fan** [comment]
ok to test

**Comment by rxin** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -713,13 +713,40 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+          case "file" => AddFile(remainder(ctx.identifier).trim)
+          case "jar" => AddJar(remainder(ctx.identifier).trim)
+          case other => throw operationNotAllowed(s"ADD with resource type '$other'", ctx)
+        }
+      case SqlBaseParser.DELETE =>
+        throw operationNotAllowed(s"DELETE resources", ctx)
+      case SqlBaseParser.LIST =>
+        ctx.identifier.getText.toLowerCase match {
+          case "files" | "file" =>
+            if (remainder(ctx.identifier).trim.length > 0) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: nit: we can define a variable `val path = remainder(ctx.identifier).trim` at the very beginning.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 6]
```diff
@@ -46,3 +51,56 @@ case class AddFile(path: String) extends RunnableCommand {
...
+
+/**
+ * Return a list of file paths that are added to resources.
+ * If file paths are provided, return the ones that are added to resources.
+ */
+case class ListFilesCommand(files: Seq[String] = Seq.empty[String]) extends RunnableCommand {
+  override val output: Seq[Attribute] = {
+    val schema = StructType(
+      StructField("Results", StringType, nullable = false) :: Nil)
+    schema.toAttributes
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: how about `AttributeReference("result", StringType, nullable = false)() :: Nil`? And it will be good if we can check what's the output name hive produces.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 7]
```diff
@@ -46,3 +51,56 @@ case class AddFile(path: String) extends RunnableCommand {
...
+case class ListJarsCommand(jars: Seq[String] = Seq.empty[String]) extends RunnableCommand {
+  override val output: Seq[Attribute] = {
+    val schema = StructType(
+      StructField("Results", StringType, nullable = false) :: Nil)
+    schema.toAttributes
+  }
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val jarList = sparkSession.sparkContext.listJars()
+    if (jars.size > 0) {
+      jars.map { f =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Should also we check the URI like `ListFilesCommand`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 5 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Thanks! Yes. I will update it once the current test build finished.

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala [hunk 8]
```diff
@@ -295,9 +294,7 @@ private[hive] class SparkSQLCLIDriver extends CliDriver with Logging {
       System.exit(0)
     }
     if (tokens(0).toLowerCase(Locale.ENGLISH).equals("source") ||
-      cmd_trimmed.startsWith("!") ||
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala]: cc @liancheng to take a look at the thrift server change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 6 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Thanks! I will update to use `AttributeReference` here. By the way, I check with hive shell, list command does not show an output name. 
```
hive> list files;
/tmp/328231e3-333c-4b7b-9900-87ef634ef86e_resources/test.txt
```

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: ok then we are free the decide the name here, `result` LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 7 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: It is a bit difference between listFiles and listJars here. 
`SparkContext.addedFiles` keeps the full file path that is same as the provided one during adding, except when the file path is provided without a protocol, in which case `file:` is added. So to look up for a file resource, i need to keep the file path as close as possible. 

On the other hand, when a jar is added to the resource, `SparkContext.addedJars` contains a jar file path with different directory parent than the provided one. For example:
```
scala> spark.sql("add jar /Users/xinwu/spark/core/src/test/resources/TestUDTF.jar")
res6: org.apache.spark.sql.DataFrame = [result: int]

scala> spark.sql("list jars").show(false)
+---------------------------------------------+
|result                                       |
+---------------------------------------------+
|spark://192.168.1.234:51589/jars/TestUDTF.jar|
+---------------------------------------------+
```
So in order to look up the jar from added resources, I just need the jar file name, instead of the whole full path. This is why I don't need to check URI like I do for `listFiles`.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 9]
```diff
@@ -156,6 +156,18 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext {
         }
         x
       }).count()
+      assert(sc.listFiles().filter(_.contains("somesuffix1")).size > 0)
```
Comment (liancheng) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Maybe:

```scala
assert(sc.listFiles().count(_.contains("somesuffix1")) == 1)
```

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 10]
```diff
@@ -156,6 +156,18 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext {
...
+      sc.stop()
+    }
+  }
+
+  test("add and list jar files") {
+    val jarPath = Thread.currentThread().getContextClassLoader.getResource("TestUDTF.jar")
+    try{
+      sc = new SparkContext(new SparkConf().setAppName("test").setMaster("local"))
+      sc.addJar(jarPath.toString)
+      assert(sc.listJars().filter(_.contains("TestUDTF.jar")).size > 0)
```
Comment (liancheng) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Same as above.

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala [hunk 11]
```diff
@@ -156,6 +156,18 @@ class SparkContextSuite extends SparkFunSuite with LocalSparkContext {
...
       }).count()
+      assert(sc.listFiles().filter(_.contains("somesuffix1")).size > 0)
+    } finally {
+      sc.stop()
+    }
+  }
+
+  test("add and list jar files") {
+    val jarPath = Thread.currentThread().getContextClassLoader.getResource("TestUDTF.jar")
+    try{
```
Comment (liancheng) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Space before `{`.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 12]
```diff
@@ -117,7 +117,7 @@ statement
         tableIdentifier partitionSpec?                                 #loadData
     | TRUNCATE TABLE tableIdentifier partitionSpec?
         (COLUMNS identifierList)?                                      #truncateTable
-    | ADD identifier .*?                                               #addResource
+    | op=(ADD | DELETE | LIST) identifier .*?                          #manageResource
```
Comment (liancheng) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Seems that `DELETE` isn't implemented in this PR. Maybe remove changes related to `DELETE` and do it in another PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 13]
```diff
@@ -774,13 +774,42 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
 
   /**
-   * Create an [[AddJarCommand]] or [[AddFileCommand]] command depending on the requested resource.
+   * Create a [[AddFileCommand]], [[AddJarCommand]], [[ListFilesCommand]] or [[ListJarsCommand]]
+   * command depending on the requested operation on resources.
+   * Expected format:
+   * {{{
+   *   ADD (FILE[s] [filepath ...] | JAR[s] [jarpath ...])
+   *   DELETE (FILE[s] [filepath ...] | JAR[s] [jarpath ...])
+   *   LIST (FILE[s] [filepath ...] | JAR[s] [jarpath ...])
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: `[S]` instead of `[s]`? Also, arguments like `filepath` and `jarpath` are required, and should be written in angle brackets (`<filepath ...>` and `<jarpath ...>`).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 14]
```diff
@@ -774,13 +774,42 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+    val mayebePaths = remainder(ctx.identifier).trim
+    ctx.op.getType match {
+      case SqlBaseParser.ADD =>
+        ctx.identifier.getText.toLowerCase match {
+          case "file" => AddFileCommand(mayebePaths)
+          case "jar" => AddJarCommand(mayebePaths)
+          case other => throw operationNotAllowed(s"ADD with resource type '$other'", ctx)
+        }
+      case SqlBaseParser.DELETE =>
+        throw operationNotAllowed(s"DELETE resources", ctx)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Same as aforementioned, let's not handle `DELETE` in this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 15]
```diff
@@ -774,13 +774,42 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+      case SqlBaseParser.LIST =>
+        ctx.identifier.getText.toLowerCase match {
+          case "files" | "file" =>
+            if (mayebePaths.length > 0) {
+              ListFilesCommand(mayebePaths.split("\\s+"))
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: So paths can't contain spaces?

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 9 above)
Comment (xwu0226) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: OK. I will change this. Thanks!

File: core/src/test/scala/org/apache/spark/SparkContextSuite.scala (hunk 11 above)
Comment (xwu0226) [inline core/src/test/scala/org/apache/spark/SparkContextSuite.scala]: Yes. Thanks! Will update.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 12 above)
Comment (xwu0226) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Yes. I was thinking that another PR will handle the implementation of it. So I add the unsupported exception in SparkSqlParser in this PR. But I will leave it to another PR for parser also. Thanks! Will update.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 16]
```diff
@@ -46,3 +51,52 @@ case class AddFileCommand(path: String) extends RunnableCommand {
     Seq.empty[Row]
   }
 }
+
+/**
+ * Return a list of file paths that are added to resources.
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Nit: "Returns" instead of "Return".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 13 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Ah, for `LIST`, paths arguments are indeed optional.

Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: For `list` command, the file path or jarpath is not required. But ADD and DELETE requires filepath or jarpath. I will update those. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 14 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: OK. Will update. thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 17]
```diff
@@ -46,3 +51,52 @@ case class AddFileCommand(path: String) extends RunnableCommand {
...
+        case f if fileList.contains(f) => f
+      }.map(Row(_))
+    } else {
+      fileList.map(Row(_))
+    }
+  }
+}
+
+/**
+ * Return a list of jar files that are added to resources.
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: "Returns" instead of "Return".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 18]
```diff
@@ -46,3 +51,52 @@ case class AddFileCommand(path: String) extends RunnableCommand {
...
+case class ListJarsCommand(jars: Seq[String] = Seq.empty[String]) extends RunnableCommand {
+  override val output: Seq[Attribute] = {
+    AttributeReference("Results", StringType, nullable = false)() :: Nil
+  }
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val jarList = sparkSession.sparkContext.listJars()
+    if (jars.size > 0) {
+      jars.map { f =>
+        new Path(f).getName
+      }.flatMap {f =>
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Space after `{`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 19]
```diff
@@ -46,3 +51,52 @@ case class AddFileCommand(path: String) extends RunnableCommand {
...
+        jarList.filter(_.contains(f))
+      }.map(Row(_))
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: What's the exact semantics of `LIST JARS <paths ...>`? According to the logic above, all added jars with the same file names instead of file paths as provided paths are listed. Namely:

```
ADD JARS /foo/bar.jar /baz/bar.jar;
LIST JARS /foo/bar.jar;
```

gives both `/foo/bar.jar` and `/baz/bar.jar`. Is this expected?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 15 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Currently, our `add file/jar` command does not support path containing space yet. For example:
```
spark-sql> add file "/Users/xinwu/spark-test/data/test 1.jar";
16/05/23 10:37:39 ERROR SparkSQLDriver: Failed in [add file "/Users/xinwu/spark-test/data/test 1.jar"]
java.net.URISyntaxException: Illegal character in path at index 0: "/Users/xinwu/spark-test/data/test 1.jar"
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3105)
	at java.net.URI$Parser.parse(URI.java:3063)
	at java.net.URI.<init>(URI.java:588)
	at org.apache.spark.SparkContext.addFile(SparkContext.scala:1404)
	at org.apache.spark.SparkContext.addFile(SparkContext.scala:1386)
	at org.apache.spark.sql.execution.command.AddFileCommand.run(resources.scala:50)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:57)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:55)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:69)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:115)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:115)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:136)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:133)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:114)
```
We can look into that with another PR. What do you think?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 16 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Thanks for catching this. Will update!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 17 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Will update!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 18 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Will do.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 19 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: `ADD JARS` command will consider jar files with the same jar file name are the same jar. For example:
```
spark-sql> add jar /Users/xinwu/spark-test/data/level1/test1.jar;
ADD JAR /Users/xinwu/spark-test/data/level1/test1.jar
Added [/Users/xinwu/spark-test/data/level1/test1.jar] to class path
Added resources: [/Users/xinwu/spark-test/data/level1/test1.jar]
0
Time taken: 1.928 seconds, Fetched 1 row(s)
spark-sql> add jar /Users/xinwu/spark-test/data/test1.jar;
ADD JAR /Users/xinwu/spark-test/data/test1.jar
Added [/Users/xinwu/spark-test/data/test1.jar] to class path
Added resources: [/Users/xinwu/spark-test/data/test1.jar]
16/05/23 11:11:12 ERROR SparkContext: Error adding jar (java.lang.IllegalArgumentException: requirement failed: JAR test1.jar already registered.), was the --addJars option used?
0
Time taken: 0.06 seconds, Fetched 1 row(s)

spark-sql> list jars;
spark://9.52.238.134:57281/jars/test1.jar, 
Time taken: 0.052 seconds, Fetched 1 row(s)
```
So it looks like the jar is added to the command directory `jars`. So for `list jars`, if I specify the full path of the source jar file, it will return empty results.

Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: By the way, currently `ADD FILE/JAR` command only can add a single file or jar at a time. There will be another PR adding the support for adding multiple files or jars at a time.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 15 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: I also just confirmed that Hive shell behaves the similarly. 
```
hive> add jar 'file:/test/test 1.jar';
Illegal character in scheme name at index 0: 'file:/test/test
Query returned non-zero code: 1, cause: java.net.URISyntaxException: Illegal character in scheme name at index 0: 'file:/test/test
hive> 
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 19 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: That makes sense, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala [hunk 20]
```diff
@@ -46,3 +51,52 @@ case class AddFileCommand(path: String) extends RunnableCommand {
...
+    } else {
+      jarList.map(Row(_))
+    }
+  }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: I wonder whether the following form is more readable...

```scala
if (jars.nonEmpty) {
  for {
    jarName <- jars.map(f => new Path(f).getName)
    jarPath <- jarList if jarPath.contains(jarName)
  } yield Row(jarPath)
} else {
  jarList.map(Row(_))
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 15 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Yea, I also checked Hive code. I think it's fine to keep it like this.

Personally I'd like to use quoted string literals for resource paths in the parser rules, so that we can handle special characters like spaces via escaping. However, unfortunately this breaks backwards compatibility.

**Comment by liancheng** [comment]
In general looks good except for a few minor comments. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 15 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala (hunk 20 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/resources.scala]: Yes. this looks nicer. I will update to use this format. Thanks!

File: sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala (hunk 8 above)
Comment (liancheng) [inline sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala]: This looks good.

**Comment by liancheng** [comment]
LGTM pending Jenkins.

**Comment by xwu0226** [comment]
@liancheng Thank you very much for your time!!

**Comment by cloud-fan** [comment]
LGTM

**Comment by liancheng** [comment]
Merging to master and branch-2.0.

**Comment by xwu0226** [comment]
@liancheng @cloud-fan @rxin Thank you very much!

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 21]
```diff
@@ -238,4 +238,23 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
+
+  test("list files") {
+    val dataFilePath = Thread.currentThread().getContextClassLoader
+      .getResource("data/files/small_kv.txt")
+    runCliWithin(2.minute)(
+      s"ADD FILE $dataFilePath" -> "",
+      s"LIST FILES" -> "small_kv.txt",
+      s"LIST FILE $dataFilePath" -> "small_kv.txt"
+    )
+  }
```
Comment (yhuai) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Seems it is failing? Can you take a look? https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-maven-hadoop-2.2/1102/testReport/junit/org.apache.spark.sql.hive.thriftserver/CliSuite/list_files/

Comment (xwu0226) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: let me check

Comment (xwu0226) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: @yhuai Should I remove the failing test case to allow the merge build test going while I continue investigate the root cause? Thanks!

Comment (xwu0226) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: The strange thing is these jenkins jobs pass:
https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-sbt-hadoop-2.6/
https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-sbt-hadoop-2.4/
https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-sbt-hadoop-2.2/
https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-sbt-hadoop-2.7/
https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-sbt-hadoop-2.3/

Anyone knows what are the difference between these successful jenkins jobs and others? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 15 above)
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: mayebePaths -> maybePaths

Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: @tedyu thanks for catching this. I will update it in my next PR, then.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 21 above)
Comment (yhuai) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Seems it fails all maven builds.
